cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
+      )
+    (txMetadata, txAuxScripts) =
+      case auxData of
+        Nothing -> (TxMetadataNone, TxAuxScriptsNone)
+        Just s ->
+          let (ms, ss) = fromMaryAuxiliaryData s
+          in  ( if null ms then
+                  TxMetadataNone
+                else
+                  TxMetadataInEra TxMetadataInMaryEra (TxMetadata ms)
+              , case ss of
+                  [] -> TxAuxScriptsNone
+                  _  -> TxAuxScripts AuxScriptsInMaryEra ss
+              )
+    txWithdrawals
+      | null (Shelley.unWdrl withdrawals) = TxWithdrawalsNone
+      | otherwise =
+          TxWithdrawals WithdrawalsInMaryEra $ fromShelleyWithdrawal withdrawals
+    txCertificates
+      | null certificates = TxCertificatesNone
+      | otherwise =
+          TxCertificates
+            CertificatesInMaryEra
+            (map fromShelleyCertificate $ toList certificates)
+            ViewTx
+    txUpdateProposal =
+      case update of
+        SNothing -> TxUpdateProposalNone
+        SJust p ->
+          TxUpdateProposal UpdateProposalInMaryEra $ fromShelleyUpdate p
+    txMintValue
+      | isZero mint = TxMintNone
+      | otherwise = TxMintValue MultiAssetInMaryEra (fromMaryValue mint) ViewTx
+
+
+toShelleyWithdrawal :: [(StakeAddress, Lovelace, build)]
+                    -> Shelley.Wdrl StandardCrypto
 toShelleyWithdrawal withdrawals =
     Shelley.Wdrl $
       Map.fromList
         [ (toShelleyStakeAddr stakeAddr, toShelleyLovelace value)
         | (stakeAddr, value, _) <- withdrawals ]
 
+
+fromShelleyWithdrawal ::  Shelley.Wdrl StandardCrypto
+                      ->  [ ( StakeAddress
+                            , Lovelace
+                            , BuildTxWith ViewTx (Witness WitCtxStake era)
+                            )
+                          ]
+fromShelleyWithdrawal (Shelley.Wdrl withdrawals) =
+  [ (fromShelleyStakeAddr stakeAddr, fromShelleyLovelace value, ViewTx)
+  | (stakeAddr, value) <- Map.assocs withdrawals
+  ]
+
+
 -- | In the Shelley era the auxiliary data consists only of the tx metadata
 toShelleyAuxiliaryData :: Map Word64 TxMetadataValue
                        -> Ledger.AuxiliaryData StandardShelley
 toShelleyAuxiliaryData m =
     Shelley.Metadata
       (toShelleyMetadata m)
 
+
+fromShelleyAuxiliaryData  :: Ledger.AuxiliaryData StandardShelley
+                          -> Map Word64 TxMetadataValue
+fromShelleyAuxiliaryData (Shelley.Metadata m) = fromShelleyMetadata m
+
+
 -- | In the Allegra and Mary eras the auxiliary data consists of the tx metadata
 -- and the axiliary scripts.
 --
-toAllegraAuxiliaryData :: forall era ledgeera.
-                          ShelleyLedgerEra era ~ ledgeera
-                       => Ledger.AuxiliaryData ledgeera ~ Allegra.AuxiliaryData ledgeera
-                       => Ledger.AnnotatedData (Ledger.Script ledgeera)
-                       => Ord (Ledger.Script ledgeera)
+toAllegraAuxiliaryData :: forall era ledgerEra.
```

Comment:
Lets follow the pattern already established in cardano-api. 

`ledgerEra` should be `ledgerera`

Reply:
done